pkg/sds/server/reconciler.go

```diff
@@ -141,3 +149,34 @@ func (d *DataplaneReconciler) generateSnapshot(dataplane *mesh_core.DataplaneRes
 	})
 	return snap, nil
 }
+
+func (d *DataplaneReconciler) updateInsights(dataplaneId core_model.ResourceKey, snapshot envoy_cache.Snapshot) error {
+	secret := snapshot.Resources[envoy_types.Secret].Items[IdentityCertResource].(*envoy_auth.Secret)
+	certPEM := secret.GetTlsCertificate().CertificateChain.GetInlineBytes()
+	block, _ := pem.Decode(certPEM)
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return err
+	}
+
+	create := false
```

Comment:
From here and down to the end that method reminds `dataplaneInsightStore.Upsert`. Maybe it makes sense to extract code to the kind of `CreateOrUpdate` methods for dataplaneInsights. What do you think?

Reply:
I think such logic should be placed in `DataplaneInsightManager` which is pending in your PR. I'll merge it for now and refactor once your stuff is merged